service/worker/scanner/history/scavenger.go

```diff
@@ -317,6 +335,46 @@ func (s *Scavenger) getPaginationFn(
 	}
 }
 
+func (s *Scavenger) cleanUpWorkflowPassedRetention(
+	ctx context.Context,
+	mutableState *persistencepb.WorkflowMutableState,
+) error {
+	executionInfo := mutableState.GetExecutionInfo()
+	ns, err := s.registry.GetNamespaceByID(namespace.ID(executionInfo.GetNamespaceId()))
+	switch err.(type) {
+	case *serviceerror.NamespaceNotFound:
+		// TODO delete the workflow data after issue #3536 close
+		return nil
+	case nil:
+		// continue to delete
+	default:
+		return err
+	}
+
+	retention := ns.Retention()
```

Comment:
What if this retention is zero? I understand that means something else went wrong because it shouldn't be, but it would lead to us deleting virtually every execution. I think we should log an error and skip in this case.

Reply:
This is no expected and it has a config executionDataDurationBuffer to protect it.